python/servo/visual_studio.py

```diff
@@ -12,80 +12,141 @@
 
-    vc_install_path = os.environ.get("VCINSTALLDIR", os.path.join(vsinstalldir, "VC"))
-    if not os.path.exists(vc_install_path):
-        print(f"Can't find Visual C++ {version_number} installation at {vc_install_path}")
-        sys.exit(1)
+def find_compatible_msvc_with_vswhere() -> Generator[VisualStudioInstallation, None, None]:
+    """Try to find the MSVC installation with the `vswhere.exe` tool. The results
+    are sorted with newer versions first."""
 
+    vswhere = find_vswhere()
+    if not vswhere:
+        return
+
+    output = subprocess.check_output([
+        vswhere,
+        '-format', 'json',
+        '-products', '*',
+        '-requires', 'Microsoft.VisualStudio.Component.VC.Tools.x86.x64',
+        '-requires', 'Microsoft.VisualStudio.Component.Windows10SDK'
+    ]).decode(errors='ignore')
+
+    for install in json.loads(output):
+        installed_version = f"{install['installationVersion'].split('.')[0]}.0"
+        if installed_version not in COMPATIBLE_MSVC_VERSIONS.values():
+            continue
+        installation_path = install['installationPath']
+        yield VisualStudioInstallation(
+            version_number=installed_version,
+            installation_path=installation_path,
+            vc_install_path=os.path.join(installation_path, "VC")
+        )
+
+
+def find_compatible_msvc_with_path() -> Generator[VisualStudioInstallation, None, None]:
+    for program_files in [PROGRAM_FILES, PROGRAM_FILES_X86]:
+        if not program_files:
+            continue
+        for (version, version_number) in COMPATIBLE_MSVC_VERSIONS.items():
+            for edition in ["Enterprise", "Professional", "Community", "BuildTools"]:
+                installation_path = os.path.join(program_files, "Microsoft Visual Studio", version, edition)
+                if os.path.exists(installation_path):
+                    yield VisualStudioInstallation(
+                        version_number=version_number,
+                        installation_path=installation_path,
+                        vc_install_path=os.path.join(installation_path, "VC")
+                    )
+
+
+def find_compatible_msvc_with_environment_variables() -> Optional[VisualStudioInstallation]:
+    installation_path = os.environ.get('VSINSTALLDIR')
+    version_number = os.environ.get('VisualStudioVersion')
+    if not installation_path or not version_number:
+        return None
+    vc_install_path = os.environ.get("VCINSTALLDIR", os.path.join(installation_path, "VC"))
+    if not os.path.exists(installation_path) or not os.path.exists(vc_install_path):
+        return None
     return VisualStudioInstallation(
         version_number=version_number,
-        installation_path=vsinstalldir,
+        installation_path=installation_path,
         vc_install_path=vc_install_path,
     )
 
 
-def find_windows_sdk_installation_path(vs_platform: str) -> str:
+def find_msvc_installations() -> VisualStudioInstallation:
```

Comment:
Doesn't the return type need to be wrapped in `Generator` type?

Reply:
Also, does it need to be a generator? it looks like the only call site is where we convert it into a list.

Reply:
That said, maybe the code is a little cleaner if it just returns a list. I'll go ahead and do that.